### PR TITLE
Add parser errorRecovery (including StandardCompiler)

### PR DESCRIPTION
**Pull request by rocky**
### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

File: libsolidity/interface/StandardCompiler.cpp [hunk 1]
```diff
@@ -462,7 +462,12 @@ boost::variant<StandardCompiler::InputsAndSettings, Json::Value> StandardCompile
 	if (sources.empty())
 		return formatFatalError("JSONError", "No input sources specified.");
 
-	ret.errors = Json::arrayValue;
+	Json::Value const& errorRecovery = _input["errorRecovery"];
+
+	if (!errorRecovery.isBool() && !errorRecovery.isNull())
```
Comment (leonardoalt) [inline libsolidity/interface/StandardCompiler.cpp]: Is the `isNull` case maybe already covered by `checkKeys`?

Comment (leonardoalt) [reply libsolidity/interface/StandardCompiler.cpp]: Commented too early, just saw the next commit ;)

**Review by leonardoalt — DISMISSED** [review]
Please squash

File: test/libsolidity/StandardCompiler.cpp [hunk 2]
```diff
@@ -57,6 +57,9 @@ bool containsError(Json::Value const& _compilerResult, string const& _type, stri
 	return false;
 }
 
+// Note that there is a warning when compilng pre-release code which
```
Comment (ekpyron) [inline test/libsolidity/StandardCompiler.cpp]:
```suggestion
// Note that there is a warning when compiling pre-release code which
```

But I don't think we need or should have this comment here anyways (especially with a "This is perhaps why" guess in it ;-)). If it's true that this is only used for ignoring pre-release warnings, then we should create a separate issue for changing this to filter those warnings out and require no warnings otherwise in the tests - if there are other warnings we should either fix them or actually check what they are (not in this PR, though, of course, but as part of that new issue).

File: liblangutil/ParserBase.h [hunk 3]
```diff
@@ -62,10 +69,17 @@ class ParserBase
...
 	///@name Helper functions
-	/// If current token value is not _value, throw exception otherwise advance token.
+	/// If current token value is not @a _value, throw exception otherwise advance token
+	//  @a if _advance is true and error recovery is in effect.
 	void expectToken(Token _value, bool _advance = true);
+
+	/// Like expectToken but if there is an error ignores tokens until
+	/// the expected token or EOS is seen. If EOS is encountered, back up to the error point,
+	/// and throw an exception so that a higher grammar rule has an opportunity to recover.
+	void expectTokenOrConsumeUntil(Token _value, std::string const& _currentNodeName, bool _advance = true);
```
Comment (ekpyron) [inline liblangutil/ParserBase.h]: This is not part of this PR and lacks an implementation in this PR, please remove from here!

File: liblangutil/ParserBase.h [hunk 4]
```diff
@@ -62,10 +69,17 @@ class ParserBase
...
 	Token currentToken() const;
 	Token peekNextToken() const;
+	std::string tokenName(Token _token);
```
Comment (ekpyron) [inline liblangutil/ParserBase.h]: Same for this.

File: liblangutil/ParserBase.h [hunk 5]
```diff
@@ -77,16 +91,26 @@ class ParserBase
 	/// Creates a @ref ParserError and annotates it with the current position and the
 	/// given @a _description.
 	void parserError(std::string const& _description);
+	void parserError(SourceLocation const& _location, std::string const& _description);
```
Comment (ekpyron) [inline liblangutil/ParserBase.h]: Same for this.

File: liblangutil/ParserBase.h [hunk 6]
```diff
@@ -77,16 +91,26 @@ class ParserBase
...
+
+	/// Creates a @ref ParserWarning and annotates it with the current position and the
+	/// given @a _description.
+	void parserWarning(std::string const& _description);
```
Comment (ekpyron) [inline liblangutil/ParserBase.h]: Same for this.

File: liblangutil/ParserBase.h [hunk 7]
```diff
@@ -77,16 +91,26 @@ class ParserBase
...
 
 	/// Creates a @ref ParserError and annotates it with the current position and the
 	/// given @a _description. Throws the FatalError.
 	void fatalParserError(std::string const& _description);
+	void fatalParserError(SourceLocation const& _location, std::string const& _description);
```
Comment (ekpyron) [inline liblangutil/ParserBase.h]: Same for this.

File: liblangutil/ParserBase.h [hunk 8]
```diff
@@ -77,16 +91,26 @@ class ParserBase
...
 
 	std::shared_ptr<Scanner> m_scanner;
 	/// The reference to the list of errors and warning to add errors/warnings during parsing
 	ErrorReporter& m_errorReporter;
 	/// Current recursion depth during parsing.
 	size_t m_recursionDepth = 0;
+	/// True if we are in parser error recovery. Usually this means we are scanning for
+	/// a synchronization token like ';', or '}'. We use this to reduce cascaded error messages.
+	bool m_inParserRecovery = false;
```
Comment (ekpyron) [inline liblangutil/ParserBase.h]: Same for this.

**Review by ekpyron — CHANGES_REQUESTED** [review]

File: liblangutil/ParserBase.h (hunk 3 above)
Comment (rocky) [reply liblangutil/ParserBase.h]: There is a confusion here part of which is due to the back and forth commit and revert that was done. 

This doesn't make sense except in the context of error recovery. That was committed into develop and then pulled. In between I added this commit. When error recovery was reverted, then that broke this. So think of this as error recovery the full version as we requested as part of the error recovery PR.  

Given where we are now, I don't think it makes sense to have two PRs for error recovery: the first part that was in and reverted and the finishing part to include an field for StandardCompiler.

File: liblangutil/ParserBase.h (hunk 4 above)
Comment (rocky) [reply liblangutil/ParserBase.h]: see above

File: liblangutil/ParserBase.h (hunk 5 above)
Comment (rocky) [reply liblangutil/ParserBase.h]: see above

File: liblangutil/ParserBase.h (hunk 6 above)
Comment (rocky) [reply liblangutil/ParserBase.h]: see above

File: liblangutil/ParserBase.h (hunk 7 above)
Comment (rocky) [reply liblangutil/ParserBase.h]: see above

File: liblangutil/ParserBase.h (hunk 8 above)
Comment (rocky) [reply liblangutil/ParserBase.h]: see above.

File: test/libsolidity/StandardCompiler.cpp (hunk 2 above)
Comment (rocky) [reply test/libsolidity/StandardCompiler.cpp]: Comment should be removed in 38ea2bbbd  This aspect confused me, but I we can deal with that some other time and way.

**Comment by chriseth** [comment]
Please rebase - the error recovery pr has been merged.

File: libsolidity/interface/StandardCompiler.cpp [hunk 9]
```diff
@@ -576,6 +574,13 @@ boost::variant<StandardCompiler::InputsAndSettings, Json::Value> StandardCompile
 	if (auto result = checkSettingsKeys(settings))
 		return *result;
 
+	if (settings.isMember("errorRecovery"))
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: Maybe
```suggestion
	if (settings.isMember("parserErrorRecovery"))
```

File: libsolidity/interface/StandardCompiler.h [hunk 10]
```diff
@@ -60,6 +60,7 @@ class StandardCompiler: boost::noncopyable
 	{
 		std::string language;
 		Json::Value errors;
+		bool errorRecovery = false;
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.h]: Maybe
```suggestion
		bool parserErrorRecovery = false;
```

**Comment by chriseth** [comment]
Still work in progress?

**Comment by rocky** [comment]
@chriseth  No longer in progress. And after this there is still #6856. But one PR at a time....

File: libsolidity/interface/StandardCompiler.cpp [hunk 11]
```diff
@@ -462,8 +462,6 @@ boost::variant<StandardCompiler::InputsAndSettings, Json::Value> StandardCompile
 	if (sources.empty())
 		return formatFatalError("JSONError", "No input sources specified.");
 
-	ret.errors = Json::arrayValue;
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]: Any reason to remove this?

File: test/libsolidity/StandardCompiler.cpp [hunk 12]
```diff
@@ -226,6 +225,44 @@ BOOST_AUTO_TEST_CASE(smoke_test)
...
+			"empty": {
+				"content": ""
+			}
+		}
+	}
+	)";
+
+	result = compile(input);
+	BOOST_CHECK(containsAtMostWarnings(result));
+
```
Comment (chriseth) [inline test/libsolidity/StandardCompiler.cpp]:
```suggestion
```

File: libsolidity/interface/StandardCompiler.cpp (hunk 11 above)
Comment (rocky) [reply libsolidity/interface/StandardCompiler.cpp]: This was most likely a rebase error. Fixed in commit 1a7e09ab7

**Review by chriseth — APPROVED** [review]
